csharp/ql/lib/semmle/code/csharp/frameworks/EntityFramework.qll

```diff
@@ -351,67 +352,137 @@ module EntityFramework {
         dbSetProp = this.getADbSetProperty(headType) and
         this.stepRev(c1, _, head, headType, 0) and
         c1.getProperty() = dbSetProp and
-        tail = SummaryComponentStack::jump(dbSetProp.getGetter()) and
+        tail = SummaryComponentStack::return() and
         dist = 0
       )
       or
       exists(Content tailHead, SummaryComponentStack tailTail, Type tailType |
-        this.requiresComponentStackOut(tailHead, tailType, tailTail, dist - 1) and
+        this.requiresComponentStackOut(tailHead, tailType, tailTail, dist - 1, dbSetProp) and
         tail = SummaryComponentStack::push(SummaryComponent::content(tailHead), tailTail) and
         this.stepRev(tailHead, tailType, head, headType, dist)
       )
     }
+
+    pragma[nomagic]
+    string getSyntheticNames(
+      SummaryComponentStack input, SummaryComponentStack output, DbContextClassSetProperty dbSet
+    ) {
+      exists(Property mapped |
+        this = dbSet.getDbContextClass() and
+        input(this, input, mapped) and
+        output(this, output, mapped, dbSet) and
+        result = dbSet.getSyntheticName() + "#" + SummaryComponentStack::getComponentStack(output)
+      )
+    }
   }
 
-  private class DbContextSaveChanges extends EFSummarizedCallable {
+  private class DbContextClassSetProperty extends Property {
     private DbContextClass c;
 
-    DbContextSaveChanges() { this = c.getASaveChanges() }
+    DbContextClassSetProperty() { this = c.getADbSetProperty(_) }
 
-    pragma[noinline]
-    private predicate input(SummaryComponentStack input, Property mapped) {
-      exists(PropertyContent head, SummaryComponentStack tail |
-        c.requiresComponentStackIn(head, _, tail, _) and
-        head.getProperty() = mapped and
-        mapped = c.getAColumnProperty(_) and
-        input = SummaryComponentStack::push(SummaryComponent::content(head), tail)
+    /**
+     * Gets the string representation for a synthetic identifier for this.
+     */
+    string getSyntheticName() {
+      exists(string qualifier, string type, string name |
+        this.hasQualifiedName(qualifier, type, name)
+      |
+        result = getQualifiedName(qualifier, type, name)
       )
     }
 
-    pragma[noinline]
-    private predicate output(SummaryComponentStack output, Property mapped) {
-      exists(PropertyContent head, SummaryComponentStack tail |
-        c.requiresComponentStackOut(head, _, tail, _) and
-        head.getProperty() = mapped and
-        mapped = c.getAColumnProperty(_) and
-        output = SummaryComponentStack::push(SummaryComponent::content(head), tail)
+    /**
+     * Gets the context class where this is a Db set property.
+     */
+    DbContextClass getDbContextClass() { result = c }
+  }
+
+  /**
+   * Holds if `input` is a valid summary component stack for property `mapped`
+   * for the context class `c`.
+   */
+  pragma[noinline]
+  predicate input(DbContextClass c, SummaryComponentStack input, Property mapped) {
```

Comment:
Could we not move this back into the class?

Reply:
Yes, we can!